### PR TITLE
[Feature][ios] Allow to define custom task

### PR DESCRIPTION
**Pull request by tuyen-vuduc**
At the moment, the lib doesn't allow to change the name of the background task on iOS, hence only one background task is allowed.

This PR will allow us as developers to define as many tasks as we desire and with any unique names we want.

There is an extra step to let iOS know about these names in our `AppDeledgate.cs`.

```
WorkmanagerPlugin.registerTask(withIdentifier: "task-identifier")
```

Readme is updated with instructions.

### CODE REVIEW

**Comment by tuyen-vuduc** [comment]
@ened Yeah, that's right.
1. We need to put any task names we want in `info.plist`
2. We need to register in `applicationDidFinishLaunch:withOptions`

In my last two projects, we need more than one tasks with different names. And I think it'll be more natural to define the name we want, not the name hard-coded. That's why I create this PR.

File: README.md [hunk 1]
```diff
@@ -103,7 +103,29 @@ Workmanager().registerOneOffTask(
...
+```
+
+- We must also define your `task-identifier` in `info.plist` file
+
+```
+<dict>
+  ....
+	<key>BGTaskSchedulerPermittedIdentifiers</key>
+	<array>
+		<string>workmanager.background.task</string>
```
Comment (ened) [inline README.md:121]: This one is not strictly needed anymore, right?

File: README.md [hunk 2]
```diff
@@ -103,7 +103,29 @@ Workmanager().registerOneOffTask(
 );
 ```
 
-Tasks registered this way will appear in the callback dispatcher using as `Workmanager.iOSBackgroundProcessingTask`.
+**NOTE**
```
Comment (ened) [inline README.md:106]: Please add a line or two to explain the purpose of registering the tasks and make a note that this MUST be done in AppDelegate and Info.plist in order to work.

File: android/src/main/kotlin/be/tramckrijte/workmanager/WorkmanagerCallHandler.kt [hunk 3]
```diff
@@ -145,14 +145,16 @@ private object UnregisterTaskHandler : CallHandler<WorkManagerCall.CancelTask> {
...
+        try {
+            when (convertedCall) {
+                is WorkManagerCall.CancelTask.ByUniqueName -> WM.cancelByUniqueName(
+                    context,
+                    convertedCall.uniqueName
+                )
+                is WorkManagerCall.CancelTask.ByTag -> WM.cancelByTag(context, convertedCall.tag)
+                WorkManagerCall.CancelTask.All -> WM.cancelAll(context)
+            }
+        } finally {}
```
Comment (ened) [inline android/src/main/kotlin/be/tramckrijte/workmanager/WorkmanagerCallHandler.kt:157]: Please handle the errors appropriately - what could go wrong here?

File: example/ios/Runner.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -509,7 +515,7 @@
 					"$(inherited)",
 					"$(PROJECT_DIR)/Flutter",
 				);
-				INFOPLIST_FILE = Runner/Info.plist;
+				INFOPLIST_FILE = "Runner/Info-Debug.plist";
```
Comment (ened) [inline example/ios/Runner.xcodeproj/project.pbxproj:518]: I'm assuming the `-Debug.plist` was added so we can debug locally?

(it's one of those things we should maybe have in a separate PR).

File: example/lib/main.dart [hunk 5]
```diff
@@ -95,150 +95,152 @@ class _MyAppState extends State<MyApp> {
...
-                    "1-failed",
-                    failedTaskKey,
-                  );
-                },
-              ),
-              //This task runs once
-              //This wait at least 10 seconds before running
-              PlatformEnabledButton(
+                Text(
+                  "One Off Tasks (Android only)",
```
Comment (ened) [inline example/lib/main.dart:143]: Is that still the case? (Android only)

**Review by ened — CHANGES_REQUESTED** [review]
Maybe we could split this PR. 
There are changes in iOS example setup, example file itself (the new nesting due to the scrollview prevents efficient review) and the good change itself. Possible?

File: README.md (hunk 1 above)
Comment (tuyen-vuduc) [reply README.md:121]: Yeah, right, it is not strictly anymore. I just kept there as an example to follow.

File: android/src/main/kotlin/be/tramckrijte/workmanager/WorkmanagerCallHandler.kt (hunk 3 above)
Comment (tuyen-vuduc) [reply android/src/main/kotlin/be/tramckrijte/workmanager/WorkmanagerCallHandler.kt:157]: Probably, we don't need it, I will remove.

File: example/lib/main.dart (hunk 5 above)
Comment (tuyen-vuduc) [reply example/lib/main.dart:143]: You're right. We don' need any the specific button any more.

**Comment by clragon** [comment]
are there any updates on this? having multiple tasks on iOS would be great.

**Comment by ened** [comment]
@tuyen-vuduc any chance you could pick this up again?

**Comment by tuyen-vuduc** [comment]
@ened Yeah, I am about to finish this within the week.

**Comment by tuyen-vuduc** [comment]
@ened Could you check why the UT fails on CI/CD check?

**Comment by ened** [comment]
> @ened Could you check why the UT fails on CI/CD check?

Can you please install flutter 2.10 on your machine, run `flutter pub upgrade` and push the changes? I had similar warnings earlier.

**Comment by ened** [comment]
hi @tuyen-vuduc, does https://stackoverflow.com/a/58224050/375209 help with the error you currently showing in the integration test log?

File: ios/Classes/WorkmanagerPlugin.h [hunk 6]
```diff
@@ -1,4 +1,7 @@
 #import <Flutter/Flutter.h>
 
 @interface WorkmanagerPlugin : NSObject<FlutterPlugin>
+
++ (void)registerTaskWithIdentifier:(NSString *) taskIdentifier;
```
Comment (ened) [inline ios/Classes/WorkmanagerPlugin.h:5]: Please document this public API

File: example/lib/main.dart [hunk 7]
```diff
@@ -95,150 +74,119 @@ class _MyAppState extends State<MyApp> {
...
+                    child: Text("Register Periodic Task (Android)"),
+                    onPressed: Platform.isAndroid
+                        ? () {
+                            Workmanager().registerPeriodicTask(
+                              simplePeriodicTask,
+                              simplePeriodicTask,
+                              initialDelay: Duration(seconds: 10),
+                            );
+                          }
+                        : null),
```
Comment (ened) [inline example/lib/main.dart:162]: This is slightly different behavior than before. Future versions of the example will have a dialog to explain the situation about periodic tasks to the user.

File: IOS_SETUP.md [hunk 8]
```diff
@@ -32,6 +32,28 @@ Additionally, you must configure the background task identifiers. The default id
...
+	<array>
+		<string>task-identifier</string>
+  </array>
+</key>
+```
+
+**NOTE: **
+- Every custom task ID has to be registered as above
+- Please refer the sample app for reference
+- The task identifier should be prefixed with app bundle id
```
Comment (ened) [inline IOS_SETUP.md:55]: If that is the case, please can you adjust the strings in the example as well (and link to the docs suggesting the prefix)

File: IOS_SETUP.md [hunk 9]
```diff
@@ -32,6 +32,28 @@ Additionally, you must configure the background task identifiers. The default id
...
+You MUST amend your `AppDelegate.swift` and `Info.plist` file to register your task ID.
+
+- AppDelegate.swift
+```
+WorkmanagerPlugin.registerTask(withIdentifier: "task-identifier")
+```
+
+- Info.plist
+```
+<key>BGTaskSchedulerPermittedIdentifiers</key>
```
Comment (ened) [inline IOS_SETUP.md:45]: please format this a bit more consistent. Also, this is discussed in line 28 already. Please simplify.

**Review by ened — COMMENTED** [review]
Looking good. Let's get the docs right and sort the conflict. Then happy to merge.

File: ios/Classes/WorkmanagerPlugin.h (hunk 6 above)
Comment (tuyen-vuduc) [reply ios/Classes/WorkmanagerPlugin.h:5]: @ened Do you mean adding a comment to this method?

File: example/lib/main.dart (hunk 7 above)
Comment (tuyen-vuduc) [reply example/lib/main.dart:162]: @ened Should I do anything with this?

File: ios/Classes/WorkmanagerPlugin.h (hunk 6 above)
Comment (tuyen-vuduc) [reply ios/Classes/WorkmanagerPlugin.h:11]: Method doc comment was added.

**Comment by tuyen-vuduc** [comment]
@ened is it ok for merging?

**Comment by tuyen-vuduc** [comment]
@ened How are you? Are you planning to merge this PR?

**Pull request merged**